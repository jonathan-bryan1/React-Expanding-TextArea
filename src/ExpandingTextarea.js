import React, { Component } from 'react'

class ExpandingTextarea extends Component {
  componentDidMount() {
    this._adjustTextarea()
  }

  componentDidUpdate(prevProps) {
    if (prevProps.value !== this.props.value) {
      this._adjustTextarea()
    }
  }

  render() {
    const { onChange, ...rest } = this.props
    return (
      <textarea
        { ...rest }
        ref={x => this.el = x}
        onChange={ this._handleChange.bind(this) }
      />
    )
  }

  _handleChange(e) {
    const { onChange } = this.props
    onChange(e)
    this._adjustTextarea(e)
  }

  _adjustTextarea({ target = this.el } = {}) {
    target.style.height = 0
    target.style.height = `${target.scrollHeight}px`
  }
}

ExpandingTextarea.defaultProps = {
  onChange: Function.prototype
}

export default ExpandingTextarea
