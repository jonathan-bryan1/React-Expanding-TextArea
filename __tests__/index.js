import React, { createRef } from 'react'
import { fireEvent, render } from '@testing-library/react'
import '@testing-library/jest-dom/extend-expect'

import ExpandingTextarea, { getHeight, resize } from '../source'

test('renders with a minimum number of rows', () => {
  const wrapStyles = {
    boxSizing: 'border-box',
    maxWidth: '500px'
  }
  const textareaStyles = {
    borderBottom: '4px solid',
    borderTop: '4px solid',
    lineHeight: '27.2px',
    padding: '10px',
    width: '100%'
  }
  const { getByText } = render(
    <div style={wrapStyles}>
      <ExpandingTextarea
        defaultValue="Some text"
        rows="3"
        style={textareaStyles}
      />
    </div>
  )

  expect(getByText('Some text')).toHaveStyle('height: 109.6px')
})

test('calls onChange & onInput when text changes', () => {
  const onChange = jest.fn()
  const onInput = jest.fn()
  const { getByText } = render(
    <ExpandingTextarea
      onChange={onChange}
      onInput={onInput}
      value="Some text"
    />
  )

  fireEvent.input(getByText('Some text'), {
    target: { value: 'Some text is here' }
  })

  expect(onChange).toBeCalled()
  expect(onInput).toBeCalled()
})

test('getHeight: returns scroll height when no rows', () => {
  global.window.getComputedStyle = jest.fn().mockImplementation(() => ({
    borderBottomWidth: '4px',
    borderTopWidth: '4px',
    lineHeight: '27.2px',
    paddingBottom: '10px',
    paddingTop: '10px'
  }))
  const el = { scrollHeight: 129 }

  expect(getHeight(undefined, el)).toEqual(137)
})

test('getHeight: returns scroll height when larger than row height', () => {
  global.window.getComputedStyle = jest.fn().mockImplementation(() => ({
    borderBottomWidth: '4px',
    borderTopWidth: '4px',
    lineHeight: '27.2px',
    paddingBottom: '10px',
    paddingTop: '10px'
  }))
  const el = { scrollHeight: 129 }

  expect(getHeight(3, el)).toEqual(137)
})

test('getHeight: returns row height when larger than scroll height', () => {
  global.window.getComputedStyle = jest.fn().mockImplementation(() => ({
    borderBottomWidth: '4px',
    borderTopWidth: '4px',
    lineHeight: '27.2px',
    paddingBottom: '10px',
    paddingTop: '10px'
  }))
  const el = { scrollHeight: 129 }

  expect(getHeight(5, el)).toEqual(164)
})

test('resize: sets style height when present', () => {
  global.window.getComputedStyle = jest.fn().mockImplementation(() => ({
    borderBottomWidth: '4px',
    borderTopWidth: '4px',
    lineHeight: '27.2px',
    paddingBottom: '10px',
    paddingTop: '10px'
  }))

  const el = { scrollHeight: 129, style: { height: '100px' } }

  resize(undefined, null)

  expect(el.style.height).toEqual('100px')

  resize(undefined, el)

  expect(el.style.height).toEqual('137px')
})

test('accepts ref', () => {
  const ref = createRef(null)
  render(<ExpandingTextarea defaultValue="Some text" ref={ref} />)
  expect(ref.current.tagName).toEqual('TEXTAREA')
})
