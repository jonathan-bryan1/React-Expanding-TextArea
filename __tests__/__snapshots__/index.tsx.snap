// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`continues to work if no onChange and no onInput 1`] = `
<DocumentFragment>
  <textarea
    style="height: 2px; overflow-y: hidden;"
  >
    Some text
  </textarea>
</DocumentFragment>
`;
